design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+- If a BMH has multiple interfaces, CAPM3 selects one to use.
+- Writes the ref information of the selected Interface into `.spec.interfaces`
+  in the corresponding Network CR.
+- Checks the configuration status of the SwitchPort CR corresponding to the
+  selected Interface CR to ensure that the network configuration is successful,
+  and then proceed with the provisioning operation.
+
+#### BareMetal Operator
+- BMO obtains LLDP information (include ID of the port in switch and
+  information of vlan active on that port) which is collected by IPA during the
+  inspection process.
+- BMO stores the information of bare metal server’s NICs in the status of BMH
+  CR. The status of each NICs includes its name, mac address and LLDP
+  information.
+
+#### BMH CRD
+Add a field containing LLDP information in `.status.hardware.nics`.
+```yaml
+ status:
+ …
+ hardware:
+ …
+ nics:
+   // The following data is obtained by Baremetal Operator and
+   // filled with each lldp information.
+   - name: eth0
+     ...
+     lldp:
+   - name: eth1
+     ...
+```
+
+### Add new CRD and controller
+
+#### Interface CRD
+```yaml
+apiVersion: network.metal3.io/v1alpha1
+kind: Interface
+metaData:
+  name: bm0-eth0
+spec:
+  portID: 10
+  switchMacAddress: XX.XX.XX.XX
+  neighborRef:
+    - kind:
+      apiVersion:
+      namespace:
+      name:
+```
+
+#### Interface Controller
+- Watch
+  - Interface CR
+- Configuration Operations
+  - Whenever a new Interface CR is created, Interface Controller will find the
+    corresponding SwitchPort CR according to `.spec.portID` and
+    `.spec.switchMacaddress`, and write the reference of the SwitchPort CR into
+    `spec.neighborRef`.
+
+#### Network CRD
+
+```yaml
+apiVersion: network.metal3.io/v1alpha1
+kind: Network
+metaData:
+  name: sample-network
+spec:
+  networkType: physicalSwitch
+  // The VLANs available
+  vlans:
+    - id:
+    - id:
+  // The untagged VLAN ID
+  vlanId:
+  interfaces:
+    - name: bm0-eth0
+    - name: bm1-eth1
+```
+
+#### Network Controller
+- Watch
+  - Network CR
+- Configuration Operations
+  - When the network configuration is changed, finds the corresponding
+    SwitchPort CR according to `.spec.neighborRef` in each Interface CR which
+    is stored in `.spec.interfaces` in the Network CR. Then changes the spec of
+    all the SwitchPort CRs.
+  - When `.spec.interfaces` increases, finds the corresponding Interface CR,
+    and then finds the corresponding SwitchPort CR, and write configuration to
+    that SwitchPort CR.
+
+#### Switch CRD
+The Switch CR is used to store the specific information of the switch including
+its name and access details.
+```yaml
+apiVersion: switch.metal3.io/v1alpha1
+kind: Switch
+metadata:
+  name: example-switch
+spec:
+  accessDetail:
+    ip: 192.0.0.1
+    OS: fos
+    port: 22
+    // Store the username and password of the switch
+    // in the form of secret
+    accessSecret: example-switch-secret
```

Comment:
we should probably have some info that can be matched through LLDP like the actual switch name (not the CR name). Maybe we could add here also a map of the switch ports, referencing the switchport CRs maybe, and listing things like allowed configurations for that port.